internal/k8s/deviceplugin/server.go

```diff
@@ -0,0 +1,308 @@
+		}
+
+		logEntry.Infof("registering server: %s", name)
+		if err := manager.RegisterDeviceServer(s.ctx, &pluginapi.RegisterRequest{
+			Version:      pluginapi.Version,
+			Endpoint:     socket,
+			ResourceName: name,
+		}); err != nil {
+			logEntry.Errorf("error registering server: %s", name)
+			return err
+		}
+
+		if err := s.monitorKubeletRestart(manager, socket); err != nil {
+			logEntry.Warnf("error monitoring kubelet restart: %s %+v", name, err)
+		}
+	}
+	return nil
+}
+
+func respToDeviceIDs(resp *podresources.ListPodResourcesResponse) map[string][]string {
+	deviceIDs := map[string][]string{}
+	for _, pod := range resp.PodResources {
+		for _, container := range pod.Containers {
+			for _, device := range container.Devices {
+				deviceIDs[device.ResourceName] = append(deviceIDs[device.ResourceName], device.DeviceIds...)
+			}
+		}
+	}
+	return deviceIDs
+}
+
+func (s *devicePluginServer) update() {
+	select {
+	case s.updateCh <- struct{}{}:
+	default:
+	}
+}
+
+func (s *devicePluginServer) monitorKubeletRestart(manager K8sManager, socket string) error {
+	logEntry := log.Entry(s.ctx).WithField("devicePluginServer", "monitorKubeletRestart")
+
+	resetCh, err := manager.MonitorKubeletRestart(s.ctx)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		logEntry.Infof("start monitoring kubelet restart: %s", s.name)
+		defer logEntry.Infof("stop monitoring kubelet restart: %s", s.name)
+		for {
+			select {
+			case <-s.ctx.Done():
+				return
+			case _, ok := <-resetCh:
+				if !ok {
+					return
+				}
+				logEntry.Infof("re registering server: %s", s.name)
+				if err = manager.RegisterDeviceServer(s.ctx, &pluginapi.RegisterRequest{
+					Version:      pluginapi.Version,
+					Endpoint:     socket,
+					ResourceName: s.name,
+				}); err != nil {
+					logEntry.Errorf("error re registering server: %s %+v", s.name, err)
+					return
+				}
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (s *devicePluginServer) GetDevicePluginOptions(_ context.Context, _ *pluginapi.Empty) (*pluginapi.DevicePluginOptions, error) {
+	return &pluginapi.DevicePluginOptions{}, nil
+}
+
+func (s *devicePluginServer) ListAndWatch(_ *pluginapi.Empty, server pluginapi.DevicePlugin_ListAndWatchServer) error {
+	logEntry := log.Entry(s.ctx).WithField("devicePluginServer", "ListAndWatch")
+
+	for {
+		resp, err := s.resourceListerClient.List(s.ctx, new(podresources.ListPodResourcesRequest))
+		if err != nil {
+			logEntry.Errorf("resourceListerClient unavailable: %+v", err)
+			return err
+		}
+
+		s.updateDevices(s.respToDeviceIDs(resp))
+
+		if err := server.Send(s.listAndWatchResponse()); err != nil {
+			logEntry.Errorf("server unavailable: %+v", err)
+			return err
+		}
+
+		select {
+		case <-s.ctx.Done():
+			logEntry.Info("server stopped")
+			return s.ctx.Err()
+		case <-time.After(s.resourcePollTimeout):
+		case <-s.updateCh:
+		}
+	}
+}
+
+func (s *devicePluginServer) respToDeviceIDs(resp *podresources.ListPodResourcesResponse) map[string]struct{} {
```

Comment:
Could we unify with function `respToDeviceIDs` from line 121?

Reply:
Nope, they are different:
* 121: returns [resource name] -> [list of device IDs for this resource] mapping
* 206: returns set of the device IDs for the given resource name